# API Documentation

## 1. Introduction
This tutorial introduces Python code you can use to access data and metadata from UMedia.

### 1.1. Installation of Python3

In this tutorial, we will illustrate how to use Python to obtain data from [UMedia](https://umedia.lib.umn.edu/). We will use Python3 (more specifically, version 3.7.4). 

The official website of Python3 is: https://www.python.org/downloads/.

### 1.2. Required packages

Once you're ready to use Python, let's first import several required packages: 

* 'json' is a python built-in module that functions as a json encoder and decoder. 
* 'requests' is a Python module that you can use to send all kinds of HTTP requests. In this tutorial, you will learn how to use requests to send simple HTTP requests in Python. 

In [2]:
import requests
import json

## 2. Formatting the URLs

Creating URLs is an important step to use the UMedia API. We need the URLs to make the request to the website. If the API is the car, the URL is the steering wheel of the car. Without it, the API won't go in the right direction.

For more info on how to format URLs for UMedia API, see: 

https://github.com/liu00222/api_document/blob/master/json_umedia/json_umedia.md
https://github.com/liu00222/api_document/blob/master/IIIF_umedia/construct_iiif_umedia.md

## 3. Using the Packages

### 3.1. Download json from UMedia

Let's use the ```requests``` and ```json``` libraries, and their built-in functions, to download data from UMedia. We'll focus below on the "Digitizing Immigrant Letters" collection. 

Let's create a URL and store it in a variable called ```my_url```. We can also concatenate strings using the "plus operation" in Python to combine different element of the URL: 

In [3]:
base_url = "https://umedia.lib.umn.edu/search.json?"
my_filter = "facets%5Bcollection_name_s"
my_key = "%5D%5B%5D=Digitizing+Immigrant+Letters"

my_url = base_url + my_filter + my_key

Notice that the url we created above would only include the first 20 items from the search results. To access data for up to 50 items at a time through a single API call, we can add ```&rows=50``` at the end of the URL:

In [4]:
my_item_num = "&rows=50"

my_url_50 = my_url + my_item_num

To keep things simple, we will use the ```my_url``` variable here, which includes 20 items. 

Now let's use the ```get()``` function in the ```requests``` package to access the data from our ```my_url``` variable: 

In [5]:
r = requests.get(my_url)

The variable called ```r``` includes our search results as a response object, which can be difficult to read and understand. You can view the contents of the object by running ```print(r.contents)```. 

Next we can use the ```loads()``` function in ```json``` to read the result into JSON:

In [6]:
data = json.loads(r.text)

The information has been stored in the variable called ```data```, which is a Python list of Python dictionaries. 

Let's take a look at how many items are stored in the ```data``` variable. This can be done with the Python ```len()``` function:

In [7]:
print(len(data))

20


You can access an item in the data list by referring to its index: ```data[0]``` refers to the first item in the list, ```data[1]``` the second, and so on. We use a zero instead of a one for the first item in the list because Python (and many other programming languages) start indexes for lists from zero.

In [8]:
data[0]

{'id': 'p16022coll264:133',
 'object': 'https://cdm16022.contentdm.oclc.org/utils/getthumbnail/collection/p16022coll264/id/133',
 'set_spec': 'p16022coll264',
 'collection_name': 'Digitizing Immigrant Letters',
 'collection_name_s': 'Digitizing Immigrant Letters',
 'collection_description': 'The Digitizing Immigrant Letters Project aims to make available on-line letters from the IHRC Archives and other collections (private individuals, partner institutions) that were written between 1850 and 1970 both by immigrants (the so-called “America letters”) and to immigrants (“homeland letters”). Global and multilingual in scope, the project encourages scholars, teachers, students and the public to interpret letters that express emotion and intimacy among loved ones separated by migration. [Project website available at: https://ihrca.dash.umn.edu/dil/]',
 'title': 'Alessandro Sisca (Riccardo Cordiferro) - letter to Lucia Fazio, 1900-10-18',
 'title_s': 'Alessandro Sisca (Riccardo Cordiferro) - 

It's important to note that because our ```data``` list starts from the index ```data[0]``` the index will end at ```data[19]```. For a list of 20 items we can't look for ```data[20]``` because it is actually indicating the 21st item in the list (which doesn't exist)! Looking for ```data[20]``` will give us a "list index" error.

In [9]:
# Python is comfortable with this
data[19]

# Python would complain for this
data[20]

IndexError: list index out of range

The next concern would be: what specific data elements are stored in each item from the ```data``` list? 

To answer the 2nd question, we need to dive into the inner dictionary and explore what information is stored there. Each dictionary includes a list of keys and values. We can tell it's a Python dictionary instead of a Python list becuase it's contained by curly brackets: 
```
{'id': 'p16022coll264:133', 'collection_name': 'Digitizing Immigrant Letters'}
```
In the example above, the first key is 'id' and its value is 'p16022coll264:133', and the second key is 'collection_name' with a value of 'Digitizing Immigrant Letters'.

We can use the following code to store all of the keys for the dictionary in ```data[0]``` into a variable called ```keys``` and print them out using for-loop. 

In [10]:
keys = list(data[0].keys())

for i in range(len(keys)):
    print(keys[i])

id
object
set_spec
collection_name
collection_name_s
collection_description
title
title_s
title_t
title_search
title_sort
description
date_created
date_created_ss
date_created_sort
creator
creator_ss
creator_sort
contributor
contributor_ss
notes
types
format
format_name
dimensions
subject
subject_ss
language
city
state
country
continent
parent_collection
parent_collection_name
contributing_organization
contributing_organization_name
contributing_organization_name_s
contact_information
local_identifier
dls_identifier
persistent_url
local_rights
page_count
record_type
first_viewer_type
viewer_type
attachment
document_type
featured_collection_order
date_added
date_added_sort
transcription
_version_
type
collection
is_compound
parent_id
thumb_url
thumb_cdn_url


From the above results, we can see all of the keys that we can use to grab values such as title, creator, transcript, page count, type, date added, etc. We can reference these keys to take a look at the title and the transcript of the first item: 

In [11]:
print(data[0]["title"])

Alessandro Sisca (Riccardo Cordiferro) - letter to Lucia Fazio, 1900-10-18


In [12]:
print(data[0]["transcription"])

Boston, Mass., 18 Ottobre 1900 Carissima Lucia, Sai perchè t’accludo questa lettera di Falcidia? Per dimostrarti coi fatti che io ho scritto a Grella, prima pre- gandolo, [poscia?] insultando, per farti toccare con mano che io non mi ero mica messo d’accordo con lui onde abbandonarti, per farti in- fire ancora una volta palese la mia innocenza. E in tutte le nostre quistioni, assieme sempre così, sap- pilo. Ma quando ti correggerai? Quando muterai sistema? Intanto ieri scrissi a Grella un’altra lettera dicendogliene di cotte e di crude. Io ti consiglio per il tuo bene e per la mia pace di essere in av- venire un pochino più seria. Se Grella ti sembra un cafone, Fal- cidia un uomo tozzo, [Pascucci?] un delinguente, Mignone un guar- diano di porci, Marziale un ricattaro[?] ecc., ecc., ti potresti però anche sba- gliare e non dovresti esser così leggiera sul dar dei giudizii. Tanto meno avresti dovuto dire che io mi ero abbaccato[?] con Grella, e che so io. Prima di scrivere e di parlare 

To print out the titles for all 20 items we grabbed from UMedia in the ```data``` list, we can use a for-loop: 

In [13]:
for i in range(len(data)):
    print(data[i]["title"])

Alessandro Sisca (Riccardo Cordiferro) - letter to Lucia Fazio, 1900-10-18
Alexey Grebenstchikoff - Letter to George Grebenstchikoff, Summer 1927
America Letters: August Aalto - Letter to Hilma Aerila, 1906-04-27
America Letters: August Aalto - Letter to Hilma Aerila, 1907-03-30
America Letters: Bert Aalto - Letter to Hilma Aerila, 1911-08-26
America Letters: Bert Aalto - Letter to Hilma Aerila, 1911-11-14
America Letters: Bert Aalto - Letter to Hilma Aerila, 1912-01-12
America Letters: Nameless to Olga, 1959-01-23
America Letters: Saimi - Letter to her Sister, 1938-04-04
Anatoly Grebenstchikoff - Letter to George Grebenstchikoff, 1917-08-14
Anna Paikens - letter to Edward Paikens, 1956-11-25
Anna Paikens - letter to Edward Paikens, 1957-02-03
Anna Paikens - letter to Edward Paikens, 1957-02-24
Anna Paikens - letter to Edward Paikens, 1957-04-14
Anna Paikens - letter to Helena Paikens, 1957-09-27
Antonietta Petris letter, 1948-09-26
Antonietta Petris letter, 1948-09-30
Antonietta Petri

Now, we can work with our ```data``` variable to grab more information about specific items in our UMedia search results. Let's consider how to save the transcripts for all of the items in our Python list to a text file.

This function takes two arguments: the first specifies the name of the file and the second specifies the task we want to do. In our case, we want to write data to the file, so  we'll use the symbol "w+" in the second argument. For more about the "open()" function and its usage: https://docs.python.org/3/library/functions.html. 
The following line of code does three things: 

1. Creates a file object called "f"
2. Create a text file called "letters.txt" on the local computer
3. Specifies that we want to write to the text file, ```"w+"```

In [14]:
f = open("letters.txt", "w+")

After opening (or creating) the file, we can write data to it using the ```write()``` function. We can use a for-loop to iteratively write out both the transcripts and titles for each item into our letters.txt file. 

In [15]:
for i in range(len(data)):
    f.write(data[i]["title"] + "\n" + data[i]["transcription"] + "\n\n")

There should now be a file called "letters.txt" in the same directory with this code. When opening it, you will see the title along with the transcript of each item. 

Remember to close a file every time you open one. This avoid memory leak issues, which can lower your CPU performance. The code to close your f file object is:

In [16]:
f.close()

### 3.2. Dowload Objects from UMedia

In this section, we will discuss how to use the API in Python to download digital objects from UMedia. We will focus here on how to download image files. 

To download images, we can use another built-in module in Python called ```shutil```. This module allows us to perform a variety of operations, such as copying and deletion, on local files. To learn more about ```shutil``` see: 
https://docs.python.org/3/library/shutil.html

First import the tool:

In [17]:
import shutil

First we need to create a URL for the image we want to download. As an example, we will use the image from the [How to construct IIIF calls to UMedia](https://github.com/liu00222/api_document/blob/master/IIIF_umedia/construct_iiif_umedia.md) documentation. 

In [18]:
image_url = "https://cdm16022.contentdm.oclc.org/digital/iiif/p16022coll208/4833/full/full/0/default.jpg"

Next, we can use the ```requests``` package to open the url image. 

In [23]:
resp = requests.get(image_url, stream = True)

Notice that we now need to set stream to 'True' in ```get()```, which we didn't do previously when working with UMedia metadata. This is because the image information is binary and not available in the JSON format, as before. 

Next, we can open the file to store the image that we grabbed: 

In [24]:
local_file = open('local_image.jpg', 'wb')

The second argument of the ```open()``` function is ```wb``` instead of ```w+``` because we need to write binary data to the file, indicated by the 'b' in 'wb'. 

Before writing the data to a 'local_image.jpg' file that we just created, we first need to set the ```resp``` variable object to ```raw.decode_content = True```! Otherwise the size of this file will always be zero, and thus cannot be opened by any computer. 

In [25]:
resp.raw.decode_content = True

Now, we can write the image data to the file. This is done by one line of code, using the ```copyfileobj()``` function from the ```shutil``` module. Don't forget to delete the image url response object before celebrating: 

In [26]:
# write the data into the local_file
shutil.copyfileobj(resp.raw, local_file)

# remove the image response object
del resp

After running the code above, you should find that the image called "local_image.jpg" is in the same directory as the code! Now you know how to download images from UMedia. 

This is the end of the tutorial. Thank you for reading! 